glom/grouping.py

```diff
@@ -0,0 +1,288 @@
+            if result is STOP:
+                tree[keyspec] = STOP
+                continue
+            done = False  # SKIP or returning a value means we still want more vals
+            if result is not SKIP:
+                acc[key] = result
+        if done:
+            return STOP
+        return acc
+    elif type(spec) is list:
+        for valspec in spec:
+            assert type(valspec) is not dict
+            # dict inside list is not valid
+            result = recurse(valspec)
+            if result is STOP:
+                return STOP
+            if result is not SKIP:
+                acc.append(result)
+        return acc
+    raise ValueError("{} not a valid spec type for Group mode".format(type(spec)))
+
+
+class First(object):
+    """
+    holds onto the first value
+
+    >>> glom([1, 2, 3], Group(First()))
+    1
+    """
+    __slots__ = ()
+
+    def agg(self, target, tree):
+        if self not in tree:
+            tree[self] = STOP
+            return target
+        return STOP
+
+
+class Avg(object):
+    """
+    takes the numerical average of all values;
+    raises exception on non-numeric value
+
+    >>> glom([1, 2, 3], Group(Avg()))
+    2.0
+    """
+    __slots__ = ()
+
+    def agg(self, target, tree):
+        if self not in tree:
+            tree[self] = [0, 0.0]
+        tree[self][0] += target
+        tree[self][1] += 1
+        return tree[self][0] / tree[self][1]
+
+
+class Sum(object):
+    """
+    takes the sum of all values;
+    raises exception on values incompatible with addition operator
+
+    >>> glom([1, 2, 3], Group(Sum()))
+    6
+    """
+    __slots__ = ()
+
+    def agg(self, target, tree):
+        if self not in tree:
+            tree[self] = 0
+        tree[self] += target
+        return tree[self]
+
+
+class Max(object):
+    """
+    takes the maximum of all values;
+    raises exception on values that are not comparable
+
+    >>> glom([1, 2, 3], Group(Max()))
+    3
+    """
+    __slots__ = ()
+
+    def agg(self, target, tree):
+        if self not in tree:
+            tree[self] = target
+        if target > tree[self]:
+            tree[self] = target
+        return tree[self]
+
+
+class Min(object):
+    """
+    takes the minimum of all values;
+    raises exception on values that are not comparable
+
+    >>> glom([1, 2, 3], Group(Min()))
+    1
+    """
+    __slots__ = ()
+
+    def agg(self, target, tree):
+        if self not in tree:
+            tree[self] = target
+        if target < tree[self]:
+            tree[self] = target
+        return tree[self]
+
+
+class Count(object):
+    """
+    takes a count of how many values occurred
+
+    >>> glom([1, 2, 3], Group(Count()))
+    3
+    """
+    __slots__ = ()
+
+    def agg(self, target, tree):
+        if self not in tree:
+            tree[self] = 0
+        tree[self] += 1
+        return tree[self]
+
+
+# TODO: is this a good idea or better to use random.sample()?
```

Comment:
I guess this is interesting because it's streaming, but because the output isn't streaming, and it's unlikely to be accessed before the `glom()` call completes, this approach *is* kinda moot.